src/libsystemd/sd-journal/compress.h

```diff
@@ -72,19 +51,38 @@ int decompress_stream_xz(int fdf, int fdt, uint64_t max_size);
 int decompress_stream_lz4(int fdf, int fdt, uint64_t max_size);
 int decompress_stream_zstd(int fdf, int fdt, uint64_t max_size);
 
-#if DEFAULT_COMPRESSION == COMPRESSION_ZSTD
-#  define compress_stream compress_stream_zstd
+static inline int compress_blob(const void *src, uint64_t src_size,
+                                void *dst, size_t dst_alloc_size, size_t *dst_size) {
+#if DEFAULT_COMPRESSION_ZSTD
+        return compress_blob_zstd(src, src_size, dst, dst_alloc_size, dst_size);
+#elif DEFAULT_COMPRESSION_LZ4
+        return compress_blob_lz4(src, src_size, dst, dst_alloc_size, dst_size);
+#elif DEFAULT_COMPRESSION_XZ
+        return compress_blob_xz(src, src_size, dst, dst_alloc_size, dst_size);
+#else
+        return -EOPNOTSUPP;
+#endif
+}
+
```

Comment:
Another possibility would be to rewrite this as a C construct instead than as a preprocessor macro. The compiler should be able to optimize it away:
```c
static inline int compress_blob(const void *src, uint64_t src_size,
                                void *dst, size_t dst_alloc_size, size_t *dst_size) {
        switch (DEFAULT_COMPRESSION) {
        case COMPRESSION_ZSTD:
                return compress_blob_zstd(src, src_size, dst, dst_alloc_size, dst_size);
        case ...
        default:
                return -EOPNOTSUPP;
        }
}
```

Reply:
This has the problem that DEFAULT_COMPRESSION must be defined to one of the constants. So the constants must be defined, and we need to duplicate their definitions in `meson.build`. This patch removes this duplication, which means that this approach is not possible.

Reply:
Maybe I should have been more explicit: turning this into a C construct you don't need the defined in Meson, you can just set DEFAULT_COMPRESSION to a C enum label
```meson
conf.set('DEFAULT_COMPRESSION', 'COMPRESSION_ZSTD')
```
as the check is done by the compiler and not by the preprocessor.

Reply:
Nope, this doesn't work.

Reply:
I just tested to make sure, and it works. Minimal example:
```meson
project('test','c')

conf = configuration_data()
conf.set('DEFAULT_COMPRESSION', 'COMPRESSION_ZSTD')

configure_file(
  output: 'config.h',
  configuration: conf)

executable('main', 'main.c')
```
```c
#include <stdio.h>

#include "config.h"

enum setting {
  COMPRESSION_ZSTD,
  COMPRESSION_XZ,
};

int main(int argc, char **argv)
{
  switch (DEFAULT_COMPRESSION) {
  case COMPRESSION_ZSTD:
    printf("zstd\n");
    return 0;
  case COMPRESSION_XZ:
    printf("xz\n");
    return 0;
  default:
    return 1;
  }
}
```
Whether you like it or not, or whether it is preferable to go this way or another way is a different matter, but it works just fine.

Reply:
The problem is that you need to define COMPRESSION_XYZ too in meson, so that config.h is available, and I guess this PR wants to avoid that

Reply:
systemd uses `-Werror=undef`. This example will fail with an error about the line `#define DEFAULT_COMPRESSION COMPRESSION_ZSTD`, something like "COMPRESSION_ZSTD is undefined".

Reply:
Are you sure? I just extended the example above to set `-Werror=undef` and with a second source files that includes `config.h` but not the definition of the enum and I don't get any error. However, this is with clang, maybe gcc is more strict.

Reply:
Oh, OK. If the macro is never used in `#if`, but only in C code, we don't get the error. This means that all places that use `#if` have to be refactored into C. But indeed, I think this looks a bit nicer I'll push that as a separate commit.